### PR TITLE
Add documentation for container checkpointing feature (KEP 2008)

### PR DESCRIPTION
**Pull request by adrianreber**
Pull request for KEP 2008:

* https://github.com/kubernetes/enhancements/pull/1990
* https://github.com/kubernetes/enhancements/issues/2008
* https://github.com/kubernetes/kubernetes/pull/104907

### CODE REVIEW

**Review by sftim — CHANGES_REQUESTED** [review]
@adrianreber thanks for this

Given that the feature is only available by directly interacting with a kubelet via HTTP, I suggest you start by writing a reference page about how to checkpoint a container via that API (to go inside https://kubernetes.io/docs/reference/node/).

I suggest that you document the body, or perhaps query string (I haven't looked) for an HTTP request that triggers a checkpoint. You should also document the local paths that the kubelet writes to (as per the existing PR), and the _format_ of the checkpoint image.

At this stage, I wouldn't add any concept page. You could however _link_ to that reference page from https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/). You might think it'd be better to link from the container page, and maybe eventually we would. For this early feature I want readers interested in checkpointing to be clear that you have to have a Pod in order to checkpoint one of its containers.

Please _avoid_ linking to a KEP except for further reading at the end of a page. The aim of documentation is so that readers don't need to refer to the source code or the KEP.

**Comment by sftim** [comment]
An example of how we document API operations:
https://kubernetes.io/docs/reference/kubernetes-api/policy-resources/resource-quota-v1/#Operations

You don't have to match that style exactly; however, it'd be helpful if any new reference docs are broadly along those lines.

**Comment by adrianreber** [comment]
@sftim Thanks for your ideas. Will try to rework this PR.

**Comment by adrianreber** [comment]
@sftim I was thinking a bit more about your suggestions.

My understanding is that I should document the API, that make sense. Should I not include any general information about checkpointing as currently in this PR in the concept page. If I should not include it in a concept page, should it be somewhere else? Next to the API documentation or not at all?

**Comment by sftim** [comment]
> Should I not include any general information about checkpointing as currently in this PR in the concept page. If I should not include it in a concept page, should it be somewhere else? Next to the API documentation or not at all?

I recommend being a little more terse - it's OK in reference documentation to say that the actual checkpointing implementation depends on the container runtime (which I believe is true) and to define only what CRI expects from that implementation.
Eg:
- format of checkpoint archive
- the archive path that the kubelet passes to the container runtime via CRI

As an example, I'd expect that you'd keep:
> Checkpointing a container is the functionality to create a stateful copy of a running container. The stateful copy of the container can be moved to another system and after restore the container will continue to run at exactly the same point it was checkpointed. The concept of checkpointing a container and restoring a container can be seen as pausing and unpausing a container with the possibility to transfer the container to another system between pausing and unpausing.

Unless someone writes a blog about it, though, there'd be no mention yet of the term “_forensic_” in that initial reference page (related: I'd assume that it's not very useful to take such evidence to a court of law and then have to explain that the feature that generated it is still alpha). All that should change if / when the checkpointing feature is scheduled to move to beta; we want production-quality docs for beta features.

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md
Comment (sftim) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md:598]:
```suggestion
  See [Kubelet Checkpoint API](/docs/reference/node/kubelet-checkpoint-api/) for more details.
```

File: content/en/docs/reference/node/_index.md
Comment (sftim) [inline content/en/docs/reference/node/_index.md:4]: Ah, you might want to wait for https://github.com/kubernetes/website/pull/31758 to merge, and then rebase your branch @adrianreber. Although Git ought to work things out, it's cleaner to rely purely on the existing commits to _main_ here.

File: content/en/docs/reference/node/kubelet-checkpoint-api.md
Comment (sftim) [inline content/en/docs/reference/node/kubelet-checkpoint-api.md:7]:
```suggestion
```
The theme renders a page title for you.

File: content/en/docs/reference/node/kubelet-checkpoint-api.md
Comment (sftim) [inline content/en/docs/reference/node/kubelet-checkpoint-api.md:15]: OK, so now this is in place, I'll suggest wording tweaks:
```suggestion
Checkpointing a container is the functionality to create a stateful copy of a
running container. Once you have a stateful copy of a container, you could
move it to a different computer for debugging or similar purposes.

If you move the checkpointed container data to a computer that's able to restore
it, that restored container continues to run at exactly the same
point it was checkpointed. You can also inspect the saved data, provided that you
have suitable tools for doing so.
```

I don't want to give people the impression that restoring to a kubelet is supported - it isn't. And we [avoid statements about the future](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future) in our docs.

File: content/en/docs/reference/node/kubelet-checkpoint-api.md
Comment (sftim) [inline content/en/docs/reference/node/kubelet-checkpoint-api.md:21]:
```suggestion
Tell the kubelet to checkpoint a specific container from the specified Pod.
```

File: content/en/docs/reference/node/kubelet-checkpoint-api.md
Comment (sftim) [inline content/en/docs/reference/node/kubelet-checkpoint-api.md:30]:
```suggestion
The checkpoint archive is in _tar_ format, and could be listed using an implementation of
[`tar`](https://pubs.opengroup.org/onlinepubs/7908799/xcu/tar.html). The contents of the
archive depend on the underlying CRI implementation (the container runtime on that node).
```

File: content/en/docs/reference/node/kubelet-checkpoint-api.md
Comment (sftim) [inline content/en/docs/reference/node/kubelet-checkpoint-api.md:54]: What happens if
- the feature isn't enabled?
- the container runtime doesn't support checkpointing?
- the checkpoint archive already exists?
- there is no storage left?

For alpha, it might be OK to say that the kubelet can also return other statuses, and maybe put in an HTML comment to say that this should be filled out properly for beta.

File: content/en/docs/reference/node/kubelet-checkpoint-api.md
Comment (sftim) [inline content/en/docs/reference/node/kubelet-checkpoint-api.md:19]:
```suggestion
### `post` checkpoint the specified container {#post-checkpoint}
```

**Review by sftim — COMMENTED** [review]
Thanks @adrianreber 

I hope all these suggestions are useful. I haven't explained all of them in detail; feel free to query anything you're not sure about.

**Comment by sftim** [comment]
/retitle Add documentation for container checkpointing feature (KEP 2008)

**Comment by sftim** [comment]
Also: you're well ahead of the deadline! Nice work in getting reviewable changes in this early.

File: content/en/docs/reference/node/_index.md
Comment (adrianreber) [reply content/en/docs/reference/node/_index.md:4]: I will rebase once that PR is merged. I just copied from the main for now to have a working PR.

File: content/en/docs/reference/node/kubelet-checkpoint-api.md
Comment (adrianreber) [reply content/en/docs/reference/node/kubelet-checkpoint-api.md]: Added information that the checkpoint archive is overwritten and that it depends on the CRI implementation.

File: content/en/docs/reference/node/kubelet-checkpoint-api.md
Comment (adrianreber) [reply content/en/docs/reference/node/kubelet-checkpoint-api.md:28]: Added a line with a link to the Kubelet authentication/authorization reference

File: content/en/docs/reference/node/kubelet-checkpoint-api.md
Comment (adrianreber) [reply content/en/docs/reference/node/kubelet-checkpoint-api.md]: Also added more details.

**Comment by adrianreber** [comment]
Thanks for your suggestions. I tried to include all of them.

**Comment by tengqm** [comment]
/hold

@adrianreber Feel free to remove the hold when the upstream changes are merged.

File: content/en/docs/reference/node/_index.md
Comment (nate-double-u) [reply content/en/docs/reference/node/_index.md:4]: #31758 has merged in now.

Comment (adrianreber) [reply content/en/docs/reference/node/_index.md]: Thanks. git handled it automatically after a rebase. The change is gone from the PR. Marking as resolved.

**Comment by nate-double-u** [comment]
/sig node
Could someone from @kubernetes/sig-node-pr-reviews take a look at this? (or suggest someone else for a tech review?)

**Comment by nate-double-u** [comment]
Just following up, could someone from https://github.com/orgs/kubernetes/teams/sig-node-pr-reviews take a look at this? (or suggest someone else for a tech review?)

**Comment by nate-double-u** [comment]
(not sure how best to tag that group...)

**Comment by sftim** [comment]
Previews:
- [Feature Gates](https://deploy-preview-31753--kubernetes-io-vnext-staging.netlify.app/docs/reference/command-line-tools-reference/feature-gates/)
- [Kubelet Checkpoint API](https://deploy-preview-31753--kubernetes-io-vnext-staging.netlify.app/docs/reference/node/kubelet-checkpoint-api/)

**Comment by tengqm** [comment]
@sftim Why remove the hold? The upstream change https://github.com/kubernetes/kubernetes/pull/104907 is still under review. The hold was to avoid accidental merge of this PR. Am I missing something?

**Comment by sftim** [comment]
Oops. I saw that a PR was merged - but it was the KEP, not the code.
/hold

**Comment by adrianreber** [comment]
Closing as it was dropped from 1.24

**Pull request closed without merging**
